dataservice/api/diagnosis/models.py

```diff
@@ -66,3 +68,49 @@ class Diagnosis(db.Model, Base):
                                db.ForeignKey('participant.kf_id'),
                                doc='the participant who was diagnosed',
                                nullable=False)
+    biospecimen_id = db.Column(KfId(),
+                               db.ForeignKey('biospecimen.kf_id'),
+                               doc='the biospecimen which recieved the '
+                               'diagnosis',
+                               nullable=True)
+
+    @validates('biospecimen_id')
+    def validate_biospecimen(self, key, value):
+        """
+        Ensure that both the diagnosis and biospecimen
+        (referred to by biospecimen_id) have the same participant
+
+        If this is not the case then raise DatabaseValidationError
+        """
+        from dataservice.api.biospecimen.models import Biospecimen
+        from dataservice.api.errors import DatabaseValidationError
+
+        if ((self.kf_id is None and self.participant_id is None) or
```

Comment:
If `participant_id` is none, won't the non-nullable `participant_id` be invalid?
Does this get in the way of that error? Why do you need to know the `kf_id` is not `None` here?

Reply:
This check has to do with that weird behavior I was showing you with the tests randomly failing. It seems like that happens in marshmallow's deserialization code. I'm wondering if there's a timing issue with the attribute validation event? Bc I can't figure out why the tests pass sometimes but not other times and I can't find a pattern. I think deserialization initializes an empty mapped object and then populates the object's fields later. When the object is initialized, I think the attribute validation event gets called and you then run into the situation where self.kf_id and self.participant_id are both None. Having this logic in the if statement seems to fix the problem and give me consistent behavior in the tests...

I was going to try changing from using attribute validation events to both before_insert and before_update ORM events, but then decided not to, since the current logic seems to work? Plus I want to avoid using two events.